kubernetes-sigs/gcp-filestore-csi-driver
# Emit GKE component version metric

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:
This PR enables the gcp-filestore-driver container to emit 'component_version' metric. This metric will be emitted only by the controller driver gcp-filestore-driver container, when the env GKE_FILESTORECSI_VERSION is setup in the driver manifest. In GKE the env variable would be populated from metadata annotations (which is automatically populated for GKE pod). This metric will enable GroupBy queries on PDCSI component version

```
- name: GKE_FILESTORECSI_VERSION
      valueFrom:
        fieldRef:
          # Component version annotation is populated by renderer
          fieldPath: metadata.annotations['components.gke.io/component-version']
```

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Tested locally as follows:

1. deploy the driver (with a --http-endpoint=:22015 and env: -GKE_FILESTORECSI_VERSION value: )
2. Check the node where the controller is running, and ssh into the node where the controller is running

```
$ kubectl get po --all-namespaces -o wide | grep gcp-filestore-csi-controller
gcp-filestore-csi-driver   gcp-filestore-csi-controller-6d889db556-787jk                  4/4     Running                      0          11m    10.128.15.207   gke-gcp-fs-csi-cluster-default-pool-60af4f1c-rv9n   <none>           <none>


$ gcloud beta compute ssh --zone <zone> "gke-gcp-fs-csi-cluster-default-pool-60af4f1c-rv9n" --project <projectname>

$ curl http://localhost:22015/metrics
# HELP component_version [ALPHA] Metric to expose the version of the FILESTORECSI GKE component.
# TYPE component_version gauge
component_version{component_version="0.1.0"} 1

```

**Does this PR introduce a user-facing change?**:

```release-note
Emit component_version metric
```
